### PR TITLE
Implement recursive and flat copy

### LINKED ISSUE #60: option to preserve folder structure of copied folder
Currently the only way to preserve the folder structure from the sources is to specify `cwd` to the parent of the source and `parents: true`.

This is already a bit cumbersome if all you want to do is preserve the folder structure of what you are copying, but if the dest is a relative path then you must also make it absolute as well so that it isn't affected by `cwd`. Overall that results in more verbose and confusing code.

There should be an option that will preserve the folder structure of all copied sources into the destination without having the specify `cwd`.

### LINKED ISSUE #61: Support preserving folder structure (recursive)
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/cpy/issues/61)

This would fix: #30, #10, #40, #60 
(If you want to take on this issue, ensure you read those issue threads too, for context)

And see https://github.com/sindresorhus/cpy/pull/32 for a previous attempt.

This tasks also includes adding support for it in https://github.com/sindresorhus/cpy-cli.

This tasks also requires you to write a *lot* of tests.

And some inspiration from other tools:
- https://github.com/shelljs/shelljs#cpoptions-source_array-dest
- https://github.com/apexearth/copy
- https://github.com/jprichardson/node-fs-extra/blob/master/docs/copy.md
- https://github.com/mysticatea/cpx
- More: https://npms.io/search?q=copy+files

Question: Should it be recursive by default? Why? Why not? Include reasoning.

---

Note: This is not an easy issue. It requires you to have good JavaScript and Node.js experience. I expect you to do a good job. Instead of asking to many questions, present solutions. The point of a issue bounty is to give me less work, not more. 🙌

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Sponsors (Total: $120.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($120.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/cpy/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### LINKED ISSUE #84: Upgrade globby to latest version
Now I have warnings in console log:
```
warning cpy > globby > fast-glob > micromatch > snapdragon > source-map-resolve > resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
warning cpy > globby > fast-glob > micromatch > snapdragon > source-map-resolve > urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
```
These warnings already fixed in all libraries. All what you need - upgarde globby to version **11.0.3**

### PR DESCRIPTION
**Pull request by Idered**
Fixes #61 
Fixes #84
Fixes #60

- [ ] optimize and refactor
- [x] update docs
- [x] add better typescript types
- [x] write more tests
- [x] handle `rename` option
- [x] copying folders structure
- [x] flattening file structure
- [x] current tests passing on windows
- [x] current tests passing on linux

Examples:

`cpy('node_modules', 'web_modules')` outputs  `web_modules/node_modules/module1`
`node_modules` files are copied recursively and structure is preserved. 

`cpy('node_modules/**', 'web_modules')`  outputs `web_modules/module1`
`node_modules` files are copied recursively and structure is preserved. Only `node_modules` content will be copied to `web_modules`.

`cpy('node_modules/**', 'web_modules', {flat: true})`  outputs `web_modules/file1`
`node_modules` files are copied recursively and structure is not preserved. Only `node_modules` content will be copied to `web_modules`.

`cpy('node_modules/**/*.cmd', 'web_modules')`  will recreate node_modules structure but all non `cmd` files will be ommited

### CODE REVIEW

**Comment by sindresorhus** [comment]
@Idered Still interested in finishing this?

**Comment by Idered** [comment]
@sindresorhus Sure, I'll tackle it this or next week. From what I remember, it will change behavior a little - results for some patterns won't be compatible with current version - is this ok?

**Comment by sindresorhus** [comment]
Yes, I can do a major release.

**Comment by sindresorhus** [comment]
@Idered Friendly bump :)

**Comment by Idered** [comment]
I've fixed some stuff, now I just need to add docs

**Comment by sindresorhus** [comment]
Just a friendly reminder in case you forgot about this PR.

**Comment by mesqueeb** [comment]
@Idered @sindresorhus I'd love to have the `copying folders structure` feature! Any way I can help?

**Comment by sindresorhus** [comment]
Bump :)

**Comment by Idered** [comment]
@sindresorhus 👋 I've finished it. I'm skipping refactoring and optimizations. Please let me know if it's any good :)

**Comment by mesqueeb** [comment]
image
I'd be careful including unrelated commits, it's cognitive strain for the author.

**Comment by Idered** [comment]
@mesqueeb I agree with you but here is my reasoning: `ava` doesn't support node.js 8 https://github.com/avajs/ava/blob/HEAD/docs/support-statement.md and breaks on it https://github.com/Idered/cpy/runs/2807531200?check_suite_focus=true

I'll revert yarn to npm. I saw yarn.lock in project but now I looked again and it's in gitignore. I'm just interested why it's on ignore list. Keeping it(or package-lock) in repo would speed up dependency installation step in github actions.

**Comment by sindresorhus** [comment]
What's the reason for switching to `node-glob`? `node-glob` is not very actively maintained and also it will be a breaking change.

File: package.json [hunk 1]
```diff
@@ -45,15 +49,17 @@
 	"dependencies": {
 		"arrify": "^2.0.1",
 		"cp-file": "^7.0.0",
-		"globby": "^9.2.0",
-		"has-glob": "^1.0.0",
+		"glob": "^7.1.6",
 		"junk": "^3.1.0",
+		"mkdirp": "^1.0.4",
```
Comment (sindresorhus) [inline package.json:54]: This is no longer needed. You can use the recursive option in `fs.mkdir`.

File: package.json [hunk 2]
```diff
@@ -11,11 +11,15 @@
 		"url": "sindresorhus.com"
 	},
 	"engines": {
-		"node": ">=8"
+		"node": ">=10"
```
Comment (sindresorhus) [inline package.json:14]: Node.js 12 can be targeted.

**Comment by Idered** [comment]
> What's the reason for switching to `node-glob`? `node-glob` is not very actively maintained and also it will be a breaking change.

To be honest, I don't remember. I've spent a lot of time on this and probably found an edge case that could not be handled by globby. . I'll look for the exact reason, just give me some time

**Comment by Idered** [comment]
@sindresorhus I've reverted node-glob to globby. It started to work with a few fixes here and there.

File: package.json (hunk 2 above)
Comment (Idered) [reply package.json]: Done, also removed node 10 from github actions

File: package.json (hunk 1 above)
Comment (Idered) [reply package.json]: Removed

**Comment by Idered** [comment]
@sindresorhus hey, let me know if you have any feedback. I would like to tackle other issues in this and in the CLI repo

File: glob-pattern.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+'use strict';
+const glob = require('globby');
+const junk = require('junk');
+const {join, basename, isAbsolute} = require('path');
+const {existsSync, lstatSync} = require('fs');
```
Comment (sindresorhus) [inline glob-pattern.js:5]: This is not a good use for destructuring. Seeing `join()` in the middle of the code means nothing. Context is important. So I would prefer not using restructuring here. Same in other places.

**Comment by sindresorhus** [comment]
> I've reverted node-glob to globby.

You didn't update the TS types.

File: index.js [hunk 4]
```diff
@@ -1,68 +1,155 @@
...
+ * @property {string} cwd
+ * @property {boolean} overwrite
+ * @property {string|Function} rename
+ */
+
+/**
+ * @typedef {object} CopyStatus
+ * @property {number} written
+ * @property {number} percent
+ */
```
Comment (sindresorhus) [inline index.js:35]: I don't really see the point of these JSDoc definitions when we have a TypeScript definition.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
For the release notes, how would a user migrate from `cpy(['source/*.png', '!source/goat.png'], 'destination', {parents: false});` and `cpy(['source/*.png', '!source/goat.png'], 'destination', {parents: true});`?

File: readme.md [hunk 5]
```diff
@@ -22,7 +22,23 @@ $ npm install cpy
...
+	await cpy('node_modules', 'destination');
+
+	// Copy node_modules content to destination
+	await cpy('node_modules/**', 'destination');
+
+	// Copy node_modules structure but skip all files except package.json files
+	await cpy('node_modules/**/*.json', 'destination');
+
+	// Copy all png files into destination without keeping directory structure
+	await cpy('**/*.png', 'destination', {flat: true});
```
Comment (sindresorhus) [inline readme.md:40]: This should be added to index.d.ts too.

Readme and index.d.ts should be fully in sync.

**Comment by sindresorhus** [comment]
> hey, let me know if you have any feedback.

Sorry for the slow response. It took me a while to get through my notifications...

**Review by martinpe36 — CHANGES_REQUESTED** [review]
Should the version number in package.json be bumped:

"version": "8.1.2" -> "version": "8.1.3"

**Comment by hummal** [comment]
@Idered Waiting on this to be fixed. Will you adjust your pr soon?

File: index.js (hunk 4 above)
Comment (Idered) [reply index.js]: Right, I've switched to `@type {import('./index').Options}` and moved CopyStatus to index.d.ts. I'm leaving other JSDoc comments - those make sense in non TS file

File: glob-pattern.js (hunk 3 above)
Comment (Idered) [reply glob-pattern.js]: Good point. I've switched to default import

**Comment by sindresorhus** [comment]
This looks good now. Thank you for working on this, @Idered.

Do you plan more PRs here or should I do a new release?

**Comment by sindresorhus** [comment]
@Idered Did you see https://github.com/sindresorhus/cpy/pull/77#issuecomment-866787314 ?

**Comment by Josmithr** [comment]
Is there an ETA for the new release containing this?

**Comment by Idered** [comment]
@sindresorhus hey, sorry for delay. I plan to do some more work on this, please don't release it yet.

**Comment by hummal** [comment]
@Idered Not sure you saw this https://github.com/gregnb/filemanager-webpack-plugin/issues/94
Webpack plugin is waiting for cpy to support recursive copy. Since it's a while I would love to have some status updates here.

**Comment by sindresorhus** [comment]
@Idered Just wondering if you still plan to work on this as it's been a while and a release is needed.

**Comment by Idered** [comment]
Yeah, sorry. I'll work on it tonight

**Comment by millerds** [comment]
Pinging on this again . . .

**Comment by millerds** [comment]
Any chance this is going to be fixed?

**Comment by mesqueeb** [comment]
I hope nothing happened to @Idered 😰

**Comment by sindresorhus** [comment]
This is continued in https://github.com/sindresorhus/cpy/pull/92.

**Comment by sindresorhus** [comment]
If anyone wants to help out, fix the things that needs to be addressed in https://github.com/sindresorhus/cpy/pull/92 and do a new pull request.

**Comment by sindresorhus** [comment]
This is now released: https://github.com/sindresorhus/cpy/releases/tag/v9.0.0

**Pull request merged**